File: springboot/src/main/java/com/mcp/server/tools/WeatherTool.java
package com.mcp.server.tools;

import com.mcp.server.service.WeatherService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

import java.util.Map;

@Component
@RequiredArgsConstructor
public class WeatherTool {

    private final WeatherService weatherService;

    public String getName() {
        return "get_weather";
    }

    public String getDescription() {
        return "Get current weather information for a specific city.";
    }

    public Map<String, Object> getInputSchema() {
        return Map.of(
            "type", "object",
            "properties", Map.of(
                "city", Map.of(
                    "type", "string",
                    "description", "Name of the city (e.g., 'Seoul', 'London')"
                )
            ),
            "required", new String[]{"city"}
        );
    }

    public Mono<String> execute(Map<String, Object> arguments) {
        String city = (String) arguments.get("city");
        
        return weatherService.getWeatherData(city)
            .map(data -> {
                Map<String, Object> weather = ((java.util.List<Map<String, Object>>) data.get("weather")).get(0);
                Map<String, Object> main = (Map<String, Object>) data.get("main");
                
                String description = (String) weather.get("description");
                Object temp = main.get("temp");
                Object humidity = main.get("humidity");
                
                return String.format("Weather in %s: %s, Temperature: %s°C, Humidity: %s%%",
                    city, description, temp, humidity);
            })
            .onErrorResume(e -> Mono.just("Error fetching weather: " + e.getMessage()));
    }
}

File: python/src/tools/weather_tool.py
from typing import Any, Dict
from src.services.weather_service import weather_service
from src.tool_registry import register_tool

@register_tool(
    name="get_weather",
    description="Get current weather information for a specific city.",
    parameters={
        "type": "object",
        "properties": {
            "city": {
                "type": "string",
                "description": "Name of the city (e.g., 'Seoul', 'London')"
            }
        },
        "required": ["city"]
    }
)
async def get_weather(city: str) -> str:
    try:
        data = await weather_service.get_weather_data(city)
        
        weather_desc = data["weather"][0]["description"]
        temp = data["main"]["temp"]
        humidity = data["main"]["humidity"]
        
        return f"Weather in {city}: {weather_desc}, Temperature: {temp}°C, Humidity: {humidity}%"
    except ValueError as e:
        return f"Error: {str(e)}"
    except Exception as e:
        return f"Error fetching weather: {str(e)}"
